tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -469,35 +475,35 @@ class SPIRVEmitter : public ASTConsumer {
       // case <literal_integer>: {...; break;}
       const Expr *caseExpr = caseStmt->getLHS();
       assert(caseExpr);
-      const auto *castExpr = dyn_cast<CastExpr>(caseExpr);
-      assert(castExpr);
-      assert(castExpr->getCastKind() == CastKind::CK_IntegralCast);
-      llvm::APSInt intValue;
-      assert(castExpr->EvaluateAsInt(intValue, astContext,
-                                     Expr::SE_NoSideEffects));
-      const int64_t value = intValue.getSExtValue();
+      assert(caseExpr->isEvaluatable(astContext));
+      Expr::EvalResult evalResult;
+      caseExpr->EvaluateAsRValue(evalResult, astContext);
```

Comment:
Should give an unimplemented error if `astContext.getIntWidth(caseExpr->getType())` is not 32 instead of directly casting `value` to `uint32_t`.

Also maybe should assert `evalResult.HasSideEffects` is false. I assume the semantic analysis can guarantee that there are no side effects when evaluating case values since what can appear as case values are fairly restricted. It cannot be any kinds expressions.